sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/PasswordPlaintextFastHashingCheck.kt

```diff
@@ -0,0 +1,128 @@
+    FunMatcher {
+        qualifier = "javax.crypto.spec.PBEKeySpec"
+        matchConstructor = true
+        withArguments("kotlin.CharArray", "kotlin.ByteArray", "kotlin.Int")
+    },
+    FunMatcher {
+        qualifier = "javax.crypto.spec.PBEKeySpec"
+        matchConstructor = true
+        withArguments("kotlin.CharArray", "kotlin.ByteArray", "kotlin.Int", "kotlin.Int")
+    },
+)
+
+@Rule(key = "5344")
+class PasswordPlaintextFastHashingCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = listOf(generateSecretFunMatcher)
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: KaFunctionCall<*>,
+        kotlinFileContext: KotlinFileContext,
+    ) = withKaSession {
+        generateSecretFunMatcher.matches(resolvedCall) || return@withKaSession
+        val secretKeyFactoryCall = callExpression.predictReceiverExpression()
+            ?.resolveToCall()
+            ?.successfulFunctionCallOrNull()
+            ?.takeIf { getInstanceFunMatcher.matches(it) }
+            ?: return@withKaSession
+        val keySpecCall = resolvedCall.argumentMapping.keys.single().predictRuntimeValueExpression()
+            .resolveToCall()
+            ?.successfulConstructorCallOrNull()
+            ?.takeIf { call -> pbeKeySpecConstructorMatchers.any { matcher -> matcher.matches(call) } }
+            ?: return@withKaSession
+
+        val iterationCountExpression = keySpecCall.argumentMapping.keys.elementAtOrNull(2) ?: return@withKaSession
+        val iterationCountValueExpression = iterationCountExpression
+            .predictRuntimeValueFromParameterDefault()
+            .predictRuntimeValueExpression()
+        val iterationCount = iterationCountValueExpression.predictRuntimeIntValue() ?: return@withKaSession
+
+        val algorithmValueExpression = secretKeyFactoryCall.argumentMapping.keys.single()
+            .predictRuntimeValueFromParameterDefault()
+            .predictRuntimeValueExpression()
+        val algorithm = algorithmValueExpression.stringValue() ?: return@withKaSession
+        val minIteration = minIterations[algorithm] ?: return@withKaSession
+
+        if (iterationCount < minIteration) {
+            val secondaryLocations = mutableListOf(SecondaryLocation(kotlinFileContext.textRange(algorithmValueExpression)))
+            if (iterationCountValueExpression.textRange != iterationCountExpression.textRange) {
+                secondaryLocations.add(SecondaryLocation(kotlinFileContext.textRange(iterationCountValueExpression)))
+            }
+
+            kotlinFileContext.reportIssue(iterationCountExpression, MESSAGE.format(minIteration), secondaryLocations)
+        }
+    }
+
+    private fun KtExpression.predictRuntimeValueFromParameterDefault(): KtExpression = withKaSession {
+        resolveToCall()
+            ?.successfulVariableAccessCall()
+            ?.symbol
+            ?.psi
+            ?.let { it as? KtParameter }
```

Comment:
Is this the best way to go from a Symbol (which in this case is a `KaValueParameterSymbol`) to its `KtParameter`, in order to get its default value?
The symbol itself, when cast to `KaValueParameterSymbol`, has a `hasDefaultValue` property, but apparently no direct way to jump to the `KtExpression` of such default value.

Reply:
Interesting, I will keep that in mind! 
In the specific case at hand, I cannot see it happening, except maybe if the iteration count or the algorithm is defined in a property a base class, which resides in a different package for which we don't have sources. However, in that case, no syntactic approach would work either: if we don't have the AST, we cannot get the default value via it.
Moreover such scenario has virtually 0 probability of happening, and in the worst case we would have a False Negative. So I would conclude that the approach is safe.